File: 1. 단계별로_풀어보기/20~29/20. 백트래킹/6. 2580.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 입력 파싱
const board = inputArr.map(row => row.split(' ').map(Number))

// 남은 0의 갯수
const entireZero = board.reduce((prev, row) => (prev += row.filter(num => num === 0).length), 0)

let row = 0
let col = 0
let isFound = false

for (row = 0; row < 9; row++) {
  for (col = 0; col < 9; col++) {
    if (board[row][col] === 0) {
      recurse(row, col, 1)
      break
    }
  }

  if (col < 9) break
}

// 출력 결과 저장할 변수 (변태적으로 쓰는건 기분탓이다)
let resultStr = ''

board.forEach(row => {
  resultStr += row.join(' ') + '\n'
})

// 출력
process.stdout.write(resultStr)

function recurse(nowRow, nowCol, filledZero) {
  const isOK = Array.from({length: 10}).fill(true)

  // 같은 행에서 이미 있는 숫자 거르기
  for (let col = 0; col < 9; col++) {
    isOK[board[nowRow][col]] = false
  }

  // 같은 열에서 이미 있는 숫자 거르기
  for (let row = 0; row < 9; row++) {
    isOK[board[row][nowCol]] = false
  }

  // 같은 9분면에서 이미 있는 숫자 거르기
  const rowStart = 3 * Math.floor(nowRow / 3)
  const colStart = 3 * Math.floor(nowCol / 3)

  for (let row = rowStart; row < rowStart + 3; row++) {
    for (let col = colStart; col < colStart + 3; col++) {
      isOK[board[row][col]] = false
    }
  }

  for (let val = 1; val < 10; val++) {
    if (isOK[val]) {
      board[nowRow][nowCol] = val

      if (filledZero === entireZero) {
        isFound = true
        return
      }

      if (isFound) return

      let foundNext = false
      for (let nextRow = nowRow; nextRow < 9; nextRow++) {
        for (let nextCol = nextRow === nowRow ? nowCol + 1 : 0; nextCol < 9; nextCol++) {
          if (board[nextRow][nextCol] === 0) {
            foundNext = true
            recurse(nextRow, nextCol, filledZero + 1)
            if (isFound) {
              return
            }
            break
          }
        }
        if (foundNext) break
      }

      if (!isFound) {
        board[nowRow][nowCol] = 0
      }
    }
  }
}

File: 1. 단계별로_풀어보기/20~29/20. 백트래킹/6. 2580.py
# pypy 로 해야 돌아간다
# python 으로 하면 시간초과

import sys

input = sys.stdin.readline
output = sys.stdout.write

board = list(list(map(int, input().split())) for i in range(9))

entireZero = sum(row.count(0) for row in board)

isFound = False

def recurse(nowRow, nowCol, filledZero):
  global isFound
  
  isOK = [True] * 10
  
  for col in range(9):
    isOK[board[nowRow][col]] = False
  
  for row in range(9):
    isOK[board[row][nowCol]] = False
  
  rowStart = 3 * (nowRow // 3)
  colStart = 3 * (nowCol // 3)
  
  for row in range(rowStart, rowStart + 3):
    for col in range(colStart, colStart + 3):
      isOK[board[row][col]] = False
  
  for val in range(1, 10):
    if isOK[val]:
      board[nowRow][nowCol] = val
      
      if filledZero == entireZero:
        isFound = True
        return
      
      if isFound:
        return
      
      foundNext = False
      for nextRow in range(nowRow, 9):
        for nextCol in range(nowCol + 1 if nextRow == nowRow else 0, 9):
          if board[nextRow][nextCol] == 0:
            foundNext = True
            recurse(nextRow, nextCol, filledZero + 1)
            if isFound:
              return
            break
        if foundNext:
          break
      
      if not isFound:
        board[nowRow][nowCol] = 0

row = 0
col = 0

for row in range(9):
  for col in range(9):
    if board[row][col] == 0:
      recurse(row, col, 1)
      break
  
  if col < 9:
    break

resultStr = ''

for row in board:
  resultStr += ' '.join(map(str, row)) + '\n'

output(resultStr)